src/TimeSeries/Backend/Graphite/GraphiteBackend.php

```diff
@@ -0,0 +1,188 @@
+        if ($result === false || !$result) {
+            throw new BackendException(curl_error($ch) . " URL: " . $url . '?' . $query);
+        }
+        curl_close($ch);
+        return $result;
+    }
+
+    private function calcTimeout($until){
+        // if until is a relative time, then we want a low cache timeout
+        $now = time();
+        $timeout = GraphiteBackend::DATA_CACHE_TIMEOUT_DEFAULT;
+        if ($until->isRelative()) {
+            $t = $now;
+        } else {
+            $t = $until->getAbsoluteTime()->getTimestamp();
+        }
+        foreach (GraphiteBackend::DATA_CACHE_TIMEOUTS as $limit => $tout) {
+            if ($t >= $now - $limit) {
+                // our 'until' time falls inside this limit
+                $timeout = $tout;
+                break;
+            }
+        }
+
+        return $timeout;
+    }
+
+    /**
+     * Construct graphite query, send request, and return one TimeSeries object.
+     *
+     * @param QueryTime $from
+     * @param QueryTime $until
+     * @param $expressionJson
+     * @param int|null $maxPoints
+     * @return TimeSeries
+     * @throws BackendException|Expression\ParsingException
+     */
+    public function queryGraphite(QueryTime $from, QueryTime $until, $expressionJson, ?int $maxPoints): TimeSeries
+    {
+        // calculate cache timeout, used in graphite query
+        $timeout = $this->calcTimeout($until);
+
+        // build expressions from JSON
+        $exp = $this->expressionFactory->createFromJson($expressionJson) -> getCanonicalStr();
+        if($maxPoints == null){
+            $maxPoints = TimeSeries::DEFAULT_MAX_POINTS;
+        }
+
+        // send out acutal query
+        $result = $this->sendQuery(
+            '/render',
+            [
+                'format' => 'json-internal',
+                'target' => [$exp],
+                'from' => $from->getGraphiteTime(),
+                'until' => $until->getGraphiteTime(),
+                'cacheTimeout' => $timeout,
+                'aggFunc' => 'avg',
+                'maxDataPoints' => $maxPoints,
+            ]
+        );
+        $jsonResult = json_decode($result, true);
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new BackendException('Invalid JSON from TS backend: ' . json_last_error_msg());
+        }
+        if (!is_array($jsonResult)) {
+            throw new BackendException('Invalid response from TS backend'); }
+
+        if(count($jsonResult)!=1){
+            throw new BackendException(
+                sprintf("wrong number of timeseries response from graphite %d, expect %d", count($jsonResult), 1)
+            );
+        }
+
+        $res = $jsonResult[0];
+
+
+
+        $newSeries = new TimeSeries();
+        $from = new DateTime();
+        $from->setTimestamp((int)$res['start']);
+        $until = new DateTime();
+        $until->setTimestamp((int)$res['end']);
+        $newSeries->setDatasource($res['name']);
```

Comment:
to avoid having to pass the data source info into the backends you could leave the signals service to fill this stuff in.